src/profilers/profile-serializer.ts

```diff
@@ -15,132 +15,270 @@
-        let node = entry.node;
-        let stack = entry.stack;
-        let location = getLocation(node);
-        stack.unshift(location.id as number);
-        if (node.hitCount > 0) {
-          const sample = new perftools.profiles.Sample({
-            locationId: stack,
-            value: [node.hitCount, node.hitCount * intervalMicros]
-          });
-          samples.push(sample);
-        }
-        for (let child of node.children) {
-          entries.push({node: child, stack: stack.slice(0)});
-        }
+    loc = this.strings.push(str) - 1;
+    this.stringsMap.set(str, loc);
+    return loc;
+  }
+}
+
+/**
+ * Takes v8 profile and populates sample, location, and function fields of
+ * profile.proto.
+ *
+ * @param profile - profile.proto with empty sample, location, and function
+ * fields.
+ * @param root - root of v8 profile tree describing samples to be appended
+ * to profile.
+ * @param appendToSamples - function which converts entry to sample(s)  and
+ * appends these to end of an array of samples.
+ * @param stringTable - string table for the existing profile.
+ */
+function serialize(
+    profile: perftools.profiles.IProfile, root: ProfileNode,
+    appendToSamples: AppendEntryToSamples, stringTable: StringTable) {
+  const samples: Array<perftools.profiles.Sample> = [];
+  const locations: Array<perftools.profiles.Location> = [];
+  const functions: Array<perftools.profiles.Function> = [];
+  const locationMap: Map<number, perftools.profiles.Location> = new Map();
+  const functionMap: Map<number, perftools.profiles.Function> = new Map();
+  const functionIdMap = new Map<string, number>();
+  const locationIdMap = new Map<string, number>();
+
+  const entries: Entry[] = root.children.map((n) => ({node: n, stack: []}));
+  while (entries.length > 0) {
+    const entry = entries.pop();
+    if (entry !== undefined) {
+      const node = entry.node;
+      const stack = entry.stack;
+      const location = getLocation(node);
+      stack.unshift(location.id as number);
+      appendToSamples(entry, samples);
+      for (let child of node.children) {
+        entries.push({node: child, stack: stack.slice()});
       }
     }
   }
 
-  function getLocation(node: TimeProfileNode): perftools.profiles.Location {
-    const id = node.callUid;
-    let location = locationMap.get(id);
-    if (location !== undefined) {
-      return location;
+  profile.sample = samples;
+  profile.location = locations;
+  profile.function = functions;
+  profile.stringTable = stringTable.strings;
+
+  function getLocation(node: ProfileNode): perftools.profiles.Location {
+    const keyStr =
+        `${node.scriptId}:${node.lineNumber}:${node.columnNumber}:${node.name}`;
+    let id = locationIdMap.get(keyStr);
+    if (id !== undefined) {
+      // id is index+1, since 0 is not valid id.
+      return locations[id - 1];
     }
-    location = new perftools.profiles.Location({id: id, line: [getLine(node)]});
+    id = locations.length + 1;
```

Comment:
Nit: `id = locations.push(location) + 1;` and remove the corresponding push line below?

Reply:
I'm not sure if that would be clear, since `id` is a field in `location`.
I could change the id after creating `location` but I think that would be unclear.